### DISCUSSION TITLE
Alternate Python version error

### DISCUSSION THREAD

**Discussion by RajasGujarathi** [apache/airflow#39905] [Q&A]
Using the docker image `apache/airflow:slim-2.9.1-python3.12`.

I installed an alternate python version `3.9`.

```bash
.
.
for ADD_PYTHON_VERSION in $ALTERNATE_PYTHON_VERSIONS; do \
          wget --quiet "https://www.python.org/ftp/python/${ADD_PYTHON_VERSION}/Python-${ADD_PYTHON_VERSION}.tgz" \
          && tar xzf "Python-${ADD_PYTHON_VERSION}.tgz" \
          && cd "Python-${ADD_PYTHON_VERSION}" && ./configure --prefix=$AIRFLOW_USER_HOME_DIR/.local --enable-optimizations && make altinstall \
          && cd .. && rm -rf "Python-${ADD_PYTHON_VERSION}*" ; \
          done 
.
.

```

When using the `PythonVirtualenvOperator` in a DAG

```py
.
.
python_3_9 = PythonVirtualenvOperator(task_id='python_3_9',
                                      requirements=["colorama==0.4.0"],
                                      system_site_packages=False,
                                      python_callable=func,
                                      python_version='3.9',
                                      dag=dag)
.
.
```

The DAG fails with an error

```
{process_utils.py:183} INFO - Executing cmd: /home/***/.local/bin/python -m virtualenv /tmp/venv_4wv1u08 --python=python3.9
{process_utils.py:187} INFO - Output:
{process_utils.py:191} INFO - PermissionError: [Errno 13] Permission denied: '/root/bin'
```

Would someone be able to help?

**Suspect**
The user `airflow` has a path `/root/bin` on its $PATH but does not have access to this directory, can this be the cause of the error?

**Comment by potiuk** [comment]
You must have somewhat kip `pip` location cashed to (/root/bin) from switching to root user when you installed your python. Not sure how because you have not published the whole dockerfile of yours. The `hash -r`  somewhere in your Dockerfile should clean the cache.  Generally you should use USER directive in the dockerfile to switch users (see examples here for inspiration https://airflow.apache.org/docs/docker-stack/build.html)

**Reply by RajasGujarathi** [reply]
I am installing the alternate Python version with the `USER airflow`. DId you mean the same or something else?


```bash

.
.

USER airflow

RUN for ADD_PYTHON_VERSION in $ALTERNATE_PYTHON_VERSIONS; do \
          wget --quiet "https://www.python.org/ftp/python/${ADD_PYTHON_VERSION}/Python-${ADD_PYTHON_VERSION}.tgz" \
          && tar xzf "Python-${ADD_PYTHON_VERSION}.tgz" \
          && cd "Python-${ADD_PYTHON_VERSION}" && ./configure --prefix=$AIRFLOW_USER_HOME_DIR/.local --enable-optimizations && make altinstall \
          && cd .. && rm -rf "Python-${ADD_PYTHON_VERSION}*" ; \
          done 

# AIRFLOW_USER_HOME_DIR=/home/airflow
WORKDIR $AIRFLOW_USER_HOME_DIR

EXPOSE 8793

ENTRYPOINT ["bash", "startup.sh"]
```

**Reply by potiuk** [reply]
You use wrong entrypoint. Read the docs https://airflow.apache.org/docs/docker-stack/entrypoint.html - they explain how to add custom startup code.

**Reply by RajasGujarathi** [reply]
Not sure, I fully understood it. 😢 

> Make sure Airflow’s entrypoint is run with exec /entrypoint "${@}" as the last command in your custom entrypoint.

Will it start webserver, scheduler, worker ... or all of those 😕

**Reply by potiuk** [reply]
same as bare image. Just try it and see.

**Reply by RajasGujarathi** [reply]
Thanks @potiuk I will have a look at the ENTRYPOINT configuration. I need more time for a better understanding.

Meanwhile, User `airflow` does not have access to `/root/bin` could that be the cause, because if we modify the PATH and remove `/root/bin` then it works as expected. But obviously, we do not know the consequence of modifying the PATH.

**Reply by potiuk** [reply]
Yes. It's because - as I explained above - your `pip` is cached likely by some previous operation where it had access. `hash` is tool that allows you to clean the cache.

**Reply by RajasGujarathi** [reply]
Thanks. @potiuk 

Trying to check the pip cache location.

Attaching a screenshot from the commands executed inside the container.

<img width="1184" alt="Screenshot 2024-05-29 at 22 10 25" src="https://github.com/apache/airflow/assets/31616516/e424ddda-abd1-4890-8158-bc3175580ebe">

DAG Logs

<img width="1106" alt="Screenshot 2024-05-29 at 22 18 26" src="https://github.com/apache/airflow/assets/31616516/d66dbe57-c48b-431e-99e4-6a4a05e079c5">

**Reply by RajasGujarathi** [reply]
**Non-optimised Dockerfile** created only for simulating the behaviour

```
FROM apache/airflow:slim-2.9.1-python3.12

ARG ALTERNATE_PYTHON_VERSIONS="3.9.0"

WORKDIR /tmp

USER root 

RUN apt-get update \
    && apt-get install -y --fix-missing --no-install-recommends \
        zlib1g-dev \
        wget \
        libsasl2-dev \
        libldap2-dev \
        libssl-dev \
        gcc \
        make

USER airflow

WORKDIR $AIRFLOW_USER_HOME_DIR

RUN pip install virtualenv \
    && for ADD_PYTHON_VERSION in $ALTERNATE_PYTHON_VERSIONS; do \
    wget --quiet "https://www.python.org/ftp/python/${ADD_PYTHON_VERSION}/Python-${ADD_PYTHON_VERSION}.tgz" \
    && tar xzf "Python-${ADD_PYTHON_VERSION}.tgz" \
    && cd "Python-${ADD_PYTHON_VERSION}" && ./configure --prefix=$AIRFLOW_USER_HOME_DIR/.local --enable-optimizations && make altinstall \
    && cd .. && rm -rf "Python-${ADD_PYTHON_VERSION}*" ; \
    done
```

---

**Steps to reproduce:**

* Build the docker image
* Run the image - `docker run -it <image_name:tag> bash`
* Command to create virtualenv - `python -m virtualenv /tmp/venviplvcl3b --python=python3.9`
* Modify the PATH environment variable, remove `/root/bin`
* Try re-creating the virtualenv

**Reply by potiuk** [reply]
* have you tried hash -r ? 
* Can you propose another fix?

**Reply by RajasGujarathi** [reply]
> have you tried `hash -r` ? 

Yes, please find the findings below

![Screenshot 2024-05-30 at 10 05 06](https://github.com/apache/airflow/assets/31616516/a1b0dffb-4227-4e06-993f-baa3b03cd6e6)


> Can you propose another fix?

Cannot think of any for now, will share as soon as I can find out something.

**Reply by RajasGujarathi** [reply]
Raised #39953
